bluebird/bluebird.c

```diff
@@ -623,28 +613,81 @@ static PyObject *bluebird_bbrk(PyObject *self, PyObject *args)
     if (!PyArg_ParseTuple(args, "ill", &pid, &addr, &heap)) 
         return NULL;
 
+    struct user_regs_struct *orig_regs = NULL;
+
     if (find_syscall_exit(pid) < 0)
-        bluebird_handle_error();
+        goto error;
 
-    struct user_regs_struct *orig_regs = set_rip_local(pid, heap);
+    orig_regs = set_rip_local(pid, heap);
 
-    if (find_syscall_entrance(pid) < 0)
-        return NULL;
+    if (orig_regs == NULL)
+        goto error;
 
-    if (ptrace(PTRACE_POKEUSER, pid, ORIG_RAX * WORD, 12) < 0) {
-        bluebird_handle_error();
-        return NULL;
-    }
+    if (find_syscall_entrance(pid) < 0 ||
+        ptrace(PTRACE_POKEUSER, pid, ORIG_RAX * WORD, 12) < 0 ||
+        ptrace(PTRACE_POKEUSER, pid, RDI * WORD, addr) < 0 ||
+        reset_ip(pid, orig_regs) < 0)
+        goto error;
 
-    if (ptrace(PTRACE_POKEUSER, pid, RDI * WORD, addr) < 0) {
-        bluebird_handle_error();
-        return NULL;
-    }
+    free(orig_regs);
+
+    Py_RETURN_NONE;
+
+error:
+
+    if (orig_regs != NULL)
+        free(orig_regs);
+
+    bluebird_handle_error();
 
-    if (reset_ip(pid, orig_regs) < 0)
+    return NULL;
+}
+
+static PyObject *bluebird_bmmap(PyObject *self, PyObject *args)
+{
+    long addr, length, heap;
+    int pid, prot, flags, offset;
+
+    if (!PyArg_ParseTuple(args, "illiiil", &pid, &addr, &length, &prot,
+                                           &flags, &offset, &heap))
         return NULL;
 
+    struct user_regs_struct *orig_regs = NULL;
+
+    if (find_syscall_exit(pid) < 0)
+       goto error;
+ 
+    orig_regs = set_rip_local(pid, heap);
+
+    if (orig_regs == NULL)
+        goto error;
+
+    if (find_syscall_entrance(pid) < 0 ||
+        ptrace(PTRACE_POKEUSER, pid, RAX * WORD, 0) ||
+        ptrace(PTRACE_POKEUSER, pid, ORIG_RAX * WORD, 9) < 0 ||
+        ptrace(PTRACE_POKEUSER, pid, RDI * WORD, addr) < 0 ||
+        ptrace(PTRACE_POKEUSER, pid, RSI * WORD, length) < 0 ||
+        ptrace(PTRACE_POKEUSER, pid, RDX * WORD, prot) < 0 ||
+        ptrace(PTRACE_POKEUSER, pid, RCX * WORD, flags) < 0 ||
+        ptrace(PTRACE_POKEUSER, pid, R8 * WORD, ~0) < 0 ||
```

Comment:
Have a generic `run_syscall` function that accepts a python class object as an argument and only change registers set from there.  This will allow both of brk and mmap functions to be combined.